webpack/hashed-folder-copy-plugin/src/HashedFolderCopyPlugin.ts

```diff
@@ -316,15 +258,16 @@ export class HashedFolderCopyPlugin implements webpack.Plugin {
       const fullAssetPath: string = path.posix.join(pathPrefix, assetPath);
       if (existingAssetNames.has(fullAssetPath)) {
         const errorMessage: string = `An asset with path "${fullAssetPath}" already exists`;
-        compilation.errors.push(new Error(errorMessage));
+        compilation.errors.push(new webpack.WebpackError(errorMessage));
         return this._renderError(errorMessage);
       }
 
-      compilation.assets[fullAssetPath] = asset;
-      module.buildInfo.assets[fullAssetPath] = asset;
+      const assetSource: webpack.sources.RawSource = new webpack.sources.RawSource(asset);
+      compilation.assets[fullAssetPath] = assetSource;
```

Comment:
There are utility methods on `compilation` for adding new assets to be emitted. We should be using those instead of directly modifying `compilation.assets`.